### ISSUE TITLE
Articulations of notes in chords are uncorrectly written in MusicXML

### ISSUE THREAD

**Issue by louisbigo** [cuthbertLab/music21#909]
**music21 version**

`6.6.0`

**Problem summary**
It seems that string/fret articulations are uncorrectly written (or not written at all) in musicXML when applied to a Note which is in a Chord (although it works well for single notes). Maybe its because in case of chords, articulations are likely to concern the whole chord (but I'm not sure of that, in any case this obviously does not apply for string/fret informations which are distinct for each note of a chord).

I found that the problem can be solved by adding the following lines in the function `noteToNotations` in `m21ToXml.py`. This is really ad-hoc to string/fret articulations and I'm sure it can be corrected in a much cleaner way. I still put the code here in case it might be useful to someone, and in case it can help to improve the code properly.

```
if chordParent is not None:
    for a in n.articulations:
        if 'TechnicalIndication' in a.classes:
            if 'StringIndication' in a.classes or 'FretIndication' in a.classes:
                if mxTechnicalMark is None:
                    mxTechnicalMark = Element('technical')
                mxTechnicalMark.append(self.articulationToXmlTechnical(a))
```

**Comment by jacobtylerwalls** [comment]
https://github.com/cuthbertLab/music21/blob/60d91ddaf4a7398337181a6168c9ec45abcb93b3/music21/articulations.py#L455-L460

Speaking only for myself, I agree it would be wise to loosen this at least as far as StringIndication and FretIndication. I would try something like this:

```diff
diff --git a/music21/musicxml/m21ToXml.py b/music21/musicxml/m21ToXml.py
index 7d06c39c1..295821392 100644
--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -4006,11 +4006,12 @@ class MeasureExporter(XMLExporterBase):
                 else:
                     notations.append(mxExpression)
 
-        # apply all articulations apart from fingerings only to first note of chord
+        # allow all articulations on first note of chord
+        # else, only allow certain ones (Fingering, FretIndication, StringIndication)
         applicableArticulations = []
         fingeringNumber = 0
         for a in chordOrNote.articulations:
-            if 'Fingering' in a.classSet:
+            if not a.classSet.isdisjoint(('Fingering', 'FretIndication', 'StringIndication')):
                 if fingeringNumber == noteIndexInChord:
                     applicableArticulations.append(a)
                 fingeringNumber += 1
```

Would you be willing to try out this patch, and if it works as expected, raise a PR including a test demonstrating the expected result?

**Comment by louisbigo** [comment]
Thanks. This does not seem to work because when the note is inside a chord, `chordOrNote` is actually replaced at the beginning of the function by `chordParent` (which does not contain articulations of notes). Thus the for loop does not iterate over articulations of the notes of the chord, but over the articulations of the chord. One solution would be to iterate over `n.articulations` instead of `chordOrNote.articulations`.

Furthermore, the content of the for loop seems very specific to fingering. Therefore I still wonder if it would not make more sense to add an additional for loop over `n.articulations` which would be dedicated to String/Fret. 

I would be happy to raise a PR when we'll have found a good solution.

**Comment by jacobtylerwalls** [comment]
Thanks for trying it out. I knew it would scan the chord only; I was just hoping your articulations also existed at the level of the chord, since it seemed to be the only place music21 currently looks. This presents an interesting question -- what if the same articulation exists on the chord and on the individual note? Should it be deduplicated? Supporting the export of articulations from the individual notes of a chord (which is not supported today) opens this box of worms--but I get that we can limit the discussion to just string/fret indications for now. Is there a reason to expect people's FretIndication data will be structured like yours or more like how music21 currently expects Fingerings to be stored, i.e. on the chord? A little survey of MusicXML writers would be useful here.

I zipped over to the MNX repo and didn't find anything fleshed out about articulations there.

**Comment by louisbigo** [comment]
Thanks for this informative answer. Now I'm suspecting that the problem might be the way I'm storing FretInidcations within m21 objects ie. in notes instead of chords. However, storing fret (or fingerings) informations (that are originally read in MusicXML _notes_) in _chord_ articulations rather than in _notes_ articulations does not seem intuitive to me. I'm wondering if there is a good reason for that. Maybe it is because MusicXML stores this kind of informations exclusively in notes, therefore m21 systematically does the job to apply them to parent chords as it is more pertinent for a majority of articulations (apparently all of them except fingering and string/fret). In this case, I guess the correct option would be to modify the MusicXML parser rather than the Music XML writer. Does this looks relevant to you ?

I found the `xmlToChord` function in xmlToM21.py which looks like a good place to make the modification (see code below) but for some reason the parser does not seem to go through that function. Any further idea would be welcome.

```
        # move spanners, expressions, articulations from first note to Chord.
        # See slur in m2 of schoenberg/op19 #2
        # but move only one of each class
        # Is there anything else that should be moved???

        seenArticulations = set()
        seenExpressions = set()

        for n in sorted(notes, key=lambda x: x.pitch.ps):
            ss = n.getSpannerSites()
            # transfer all spanners from the notes to the chord.
            for sp in ss:
                sp.replaceSpannedElement(n, c)
            for art in n.articulations:
                if type(art) in seenArticulations:  # pylint: disable=unidiomatic-typecheck
                    pass
                c.articulations.append(art)
                seenArticulations.add(type(art))
            for exp in n.expressions:
                if type(exp) in seenExpressions:  # pylint: disable=unidiomatic-typecheck
                    pass
                c.expressions.append(exp)
                seenExpressions.add(type(exp))

            n.articulations = []
            n.expressions = []
```

**Comment by jacobtylerwalls** [comment]
That function indeed runs -- and at the very end you see the articulations on the notes being wiped (`n.articulations = []`) after being transferred to the chord. There's an attempt to deduplicate by only transferring one of instance of each unique subclass, but it's not working because the `pass` should be a `continue`. We're probably getting lucky based on how most MusicXML writers operate.

All this to say, I think my original patch would suffice given the current assumption music21 makes about where to store articulations. But clearly there are some todos here:

- fix parser to use `continue` instead of `pass` so that deduplication happens correctly
- but exempt certain artics from the deduplication (Fingering, FretIndication, StringIndication)
- apply something like my original suggested patch to the exporter
- add some tests, both for your original use case as well as the deduplication on input we just saw wasn't working
- add some documentation, both in this module and maybe even a short example in Ch. 28 of the User's Guide

Fun stuff, but not a nothing project. I'd gladly review a PR if you had time to put one together. Appreciate the work done so far!

**Comment by jacobtylerwalls** [comment]
I suppose there could be an additional feature -- assuming Michael is OK with the added complexity -- to scan chord members for articulations and dedupe them on export in a similar way to import, which is more like what you originally proposed. I was just trying to propose a smaller change until I stepped through the current scenario. I like the idea of having the same logic on import and export so that things Just Work.

**Comment by louisbigo** [comment]
Ok, I see the box of worms now. I am mostly interested in your last proposition (assuming Michael is OK with the added complexity) but I understand that it would be more clean to make things in the right order and to begin by the todo list in your previous message. I'll try to do at least some of these points soon (I need to solve some merge issues first).

**Comment by mscuthbert** [comment]
The problems with letting a clarinetist create your music processing library -- I never thought about multiple fingers being used in playing notes or that someone would want to specify which finger/fret/string went with each note of a chord.  Supporting this is obviously something to do.  How to do so...I'll have to think about when it's not 2am, but these approaches sound like a good approach.  I want to avoid situations like I used to get in Finale where 4-note chords would sometimes have four accent marks on them, but I think that looking at individual notes to find specific expressions (if not articulations) and then applying them to the main chord tag is probably a good approach  Round-trip testing will probably be key,   Immediately coming to mind is storing expressions on the first note of the chord instead of the chord itself and making `.articulations` and `.expressions` on a chord look at the first note for them, but people shoiuld be able to manipulate pitches in a chord without being afraid that their Fermata will disappear.  Maybe expressions should have a `.attachToNotesInChords = False` default class property and Fingering/String/etc. should override that so we know to put them on the note in a chord and not on the chord itself?

**Comment by jacobtylerwalls** [comment]
Agreed, formalizing this  in an attribute is easier to document and discover than just hard coding it in the musicxml parsers.

w/r/t making it easy to not lose your fermata, that's why we should encourage users to store artics and exprs on the chord in most cases. That's why I was +/- on the suggested feature of scanning for every articulation on every chord and deduping on export, but as long as we're deduping, it shouldn't hurt.
